### PR TITLE
Added support for `@TemplateGlobal` annotation

### LINKED ISSUE #605: Support for `@TemplateGlobal`
Support for `@TemplateGlobal`. See https://quarkus.io/guides/qute-reference#global_variables

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added support for `@TemplateGlobal` annotation.
templateglobaldemo

Fixes #605 

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	@Override
+	protected void processAnnotation(IJavaElement javaElement, IAnnotation annotation, String annotationName,
+		SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (javaElement.getElementType() != IJavaElement.TYPE) {
+			return;
+		}
+		IType type = (IType) javaElement;
+		String name = null;
+		boolean hasTemplateGlobal = false;
+		for (IAnnotation typeAnnotation : type.getAnnotations()) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:85]: I suggest that you see the @Target of @TemplateGlobal at https://github.com/quarkusio/quarkus/blob/main/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateGlobal.java#L30

It means that : 

 * you can annotate type, field and method.
 * you can annotate those java elements with only one annotation.

It means that : 

 * you must not iterate from @TemplateGlobal
 * you can have @TemplateGlobal annotation in type, field, and method. What is the behavior if type is not annotated with @TemplateGlobal but a field is annotated with @TemplateGlobal. I suggest that you test that and you provide the code which have the same behavior of Qute.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			}
+		}
+		if (!hasTemplateGlobal) {
+			return;
+		}
+
+		ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(type);
+
+		// Loop for static fields
+		IField[] fields = type.getFields();
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:101]: I suggest that you read Javadoc at https://github.com/quarkusio/quarkus/blob/5e6fe630375c96b18228f54c6537f445f3ee1743/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateGlobal.java#L10 which explains rules that field and method must match (ex : static, etc) and implement those rules.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+				typeResolver, monitor);
+		}
+	}
+
+	private void collectResolversForTemplateGlobal(IMember member, String name, List<ValueResolverInfo> resolvers,
+		ITypeResolver typeResolver, IProgressMonitor monitor) {
+		try {
+			// every non-void non-private static method that declares no parameters and
+			// every non-private static field is considered a global variable
+			if (!Modifier.isPrivate(member.getFlags()) && Modifier.isStatic(member.getFlags())) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:121]: It is not enough, I suggest that you read Javadoc at https://github.com/quarkusio/quarkus/blob/5e6fe630375c96b18228f54c6537f445f3ee1743/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateGlobal.java#L10 which explains rules that field and method must match (ex : static, etc) and implement those rules.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 2 above)
Comment (AlexXuChen) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java]: So both a field and method must be non private and static, and a method must have no parameters and return void, correct?

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 3 above)
Comment (AlexXuChen) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java]: So both a field and method must be non private and static, and a method must have no parameters and return void, correct?

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (!(javaElement instanceof IAnnotatable)) {
+			return;
+		}
+		IAnnotation templateGlobal = AnnotationUtils.getAnnotation((IAnnotatable) javaElement,
+			TEMPLATE_GLOBAL_ANNOTATION);
+		if (templateGlobal == null) {
+			return;
+		}
+		if (javaElement instanceof IType) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:86]: `  if (javaElement.getElementType() == IJavaElement.TYPE) {`

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			IType type = (IType) javaElement;
+			collectResolversForTemplateGlobal(type, templateGlobal, context.getDataModelProject().getValueResolvers(),
+				monitor);
+		} else {
+			ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver((IType) javaElement);
+
+			if (javaElement instanceof IMethod) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:93]:
```
  if (javaElement.getElementType() == IJavaElement.FIELD || javaElement.getElementType() == IJavaElement.METHOD) {
    IMember member = (IMember) javaElement;
    collectResolversForTemplateGlobal(member, context.getDataModelProject().getValueResolvers(),
					typeResolver, monitor);
```

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 3 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java]: Yes, it misses the conditions for methods (no parameters and return void)

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	private void collectResolversForTemplateGlobal(IMember member, List<ValueResolverInfo> resolvers,
+		ITypeResolver typeResolver, IProgressMonitor monitor) {
+		// every non-void non-private static method that declares no parameters and
+		// every non-private static field is considered a global variable
+		if (isTemplateGlobalMember(member)) {
+			String sourceType = member.getDeclaringType().getFullyQualifiedName();
+			ValueResolverInfo resolver = new ValueResolverInfo();
+			resolver.setSourceType(sourceType);
+			resolver.setSignature(getSignature(member, typeResolver));
+			// The name is defaulted, i.e. the name of the field/method is used.
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:135]: `resolver.setGlobalVariable(true);`

Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:135]: It seems that you can override the name of @TemplateGlobal

```
@TemplateGlobal("foo");
private String bar;
```

Your value resolver doesn't take care of that. You should get the "foo" value and if it is not empty, you should set the value to the generated resolver:

```
IAnnotation templateGlobal = AnnotationUtils.getAnnotation(member,TEMPLATE_GLOBAL_ANNOTATION);
if (templateGlobal  != null) {
    // get named from @TemplateGlobal
    String named = ...
    resolver.setNamed(named);
}
```

**Comment by AlexXuChen** [comment]
Template global example:
```java
package org.acme;

import io.quarkus.qute.TemplateGlobal;

enum Color { RED, GREEN, BLUE }

@TemplateGlobal
public class Globals {

    static int age = 40;

    static String name;

    static Color[] myColors() {
      return new Color[] { Color.RED, Color.BLUE };
    }

    @TemplateGlobal(name = "currentUser")
    static String user() {
       return "Mia";
    }
}
```

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 6 above)
Comment (AlexXuChen) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java]: Quarkus doesn't seem to support the declaration of `@TemplateGlobal("foo")`, and the compilation fails when I build the binary:
Screenshot from 2022-05-26 11-22-02
Should this be  "name" instead?

**Comment by AlexXuChen** [comment]
`name` attribute is also supported:
templateglobaldemoname

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetDataModelProjectTest.java [hunk 7]
```diff
@@ -312,6 +313,30 @@ private static void testValueResolversFromTemplateEnum(List<ValueResolverInfo> r
 			resolvers);
 	}
 
+	private static void testValueResolversFromTemplateGlobal(List<ValueResolverInfo> resolvers) {
+
+		// @TemplateGlobal
+		// public class Globals {
+
+		// static int age = 40;
+		assertValueResolver(null, "age : int", "org.acme.qute.Globals", resolvers);
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetDataModelProjectTest.java:322]: It should be nice to check globalVariable (true).

Your test should look like:

`assertValueResolver(null, "age : int", "org.acme.qute.Globals", true, resolvers);`

Inother words add the globalvariable parameter to assertValueResolver

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/datamodel/resolvers/ValueResolverInfo.java [hunk 8]
```diff
@@ -156,6 +156,7 @@ public int hashCode() {
 		result = prime * result + ((namespace == null) ? 0 : namespace.hashCode());
 		result = prime * result + ((signature == null) ? 0 : signature.hashCode());
 		result = prime * result + ((sourceType == null) ? 0 : sourceType.hashCode());
+		result = prime * result + ((globalVariable == null) ? 0 : globalVariable.hashCode());
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/datamodel/resolvers/ValueResolverInfo.java:159]: Good catch! Could you do that too in qute.ls please.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+			TEMPLATE_GLOBAL_ANNOTATION);
+		if (templateGlobal == null) {
+			return;
+		}
+		ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver((IMember) javaElement);
+		if (javaElement.getElementType() == IJavaElement.TYPE) {
+			IType type = (IType) javaElement;
+			collectResolversForTemplateGlobal(type, templateGlobal, context.getDataModelProject().getValueResolvers(),
+				typeResolver, monitor);
+		} else {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:92]:
```java
} else if (javaElement.getElementType() == IJavaElement.FIELD
				|| javaElement.getElementType() == IJavaElement.METHOD) {
...
```

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+		return ANNOTATION_NAMES;
+	}
+
+	@Override
+	protected void processAnnotation(IJavaElement javaElement, IAnnotation annotation, String annotationName,
+		SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (!(javaElement instanceof IAnnotatable)) {
+			return;
+		}
+		IAnnotation templateGlobal = AnnotationUtils.getAnnotation((IAnnotatable) javaElement,
```
Comment (rgrunber) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:82]: I think `templateGlobal` can be removed in favour of just using `annotation` here, unless I'm missing something.

They certainly match, and when I look at `AbstractAnnotationTypeReferenceDataModelProvider#collectDataModel(..)` the java element is derived from calling `getParent()` on the annotation.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	private static boolean isTemplateGlobalMember(IMember member) {
+		try {
+			// every non-void non-private static method that declares no parameters and
+			// every non-private static field is considered a global variable
+			if (!Modifier.isPrivate(member.getFlags()) && Modifier.isStatic(member.getFlags())) {
+				if (member.getElementType() == IJavaElement.FIELD) {
+					return true;
+				} else if (member.getElementType() == IJavaElement.METHOD) {
+					IMethod method = (IMethod) member;
+					return method.getParameters().length == 0 && !"void".equals(method.getReturnType());
```
Comment (rgrunber) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:179]: I would use `method.getNumberOfParameters()`

Also, try defining a void method like :

`public static void doesNothing () {}`

It shows up in the completion when the class is annotated with `@TemplateGlobal` but it shouldn't.

In fact, the same problem exists with `@TemplateExtension`. It doesn't exclude void values.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+			}
+		}
+	}
+
+	private void collectResolversForTemplateGlobal(IType type, IAnnotation templateGlobal,
+		List<ValueResolverInfo> resolvers, ITypeResolver typeResolver, IProgressMonitor monitor) {
+		try {
+			IField[] fields = type.getFields();
+			for (IField field : fields) {
+				if (AnnotationUtils.getAnnotation(field, TEMPLATE_GLOBAL_ANNOTATION) == null
```
Comment (rgrunber) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:107]: I would use `!AnnotationUtils.hasAnnotation(..)`, here and below.

**Review by rgrunber — CHANGES_REQUESTED** [review]
Works pretty well for me, but noticed a few things worth fixing.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 11 above)
Comment (AlexXuChen) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:179]: There seems to be an issue with `org.eclipse.jdt.core.Signature`, where it receives a `char[]` of the typeName:

but the resulting string is the first character:

There might be an alternative to `IMethid.getReturnType()` instead?

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java (hunk 6 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java]: Yes sorry, I meant "name" attribute.

**Comment by angelozerr** [comment]
There is compilation problem https://github.com/redhat-developer/quarkus-ls/runs/6632720210?check_suite_focus=true#step:8:142 you miss to define this method since you add globalVariable parameter (you define a new assertResolver by calling false for globalVariable).

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+		return ANNOTATION_NAMES;
+	}
+
+	@Override
+	protected void processAnnotation(IJavaElement javaElement, IAnnotation annotation, String annotationName,
+		SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (!(javaElement instanceof IAnnotatable)) {
+			return;
+		}
+//		IAnnotation tempalteGlobal = AnnotationUtils.getAnnotation((IAnnotatable) javaElement,
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:82]: remove this comments

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+			ValueResolverInfo resolver = new ValueResolverInfo();
+			resolver.setSourceType(sourceType);
+			resolver.setSignature(getSignature(member, typeResolver));
+			// Constant value for {@link #name()} indicating that the field/method name
+			// should be used
+			try {
+				resolver.setNamed(
+					AnnotationUtils.getAnnotationMemberValue(templateGlobal, TEMPLATE_GLOBAL_ANNOTATION_NAME));
+			} catch (JavaModelException e) {
+				LOGGER.log(Level.SEVERE, "Error while getting annotation member value of 'named'.", e);
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java:135]: Error while getting annotation member value of 'name'

**Comment by angelozerr** [comment]
Great job @AlexXuChen !

**Pull request merged**